File: 3133. Minimum Array End/solution_bitarray.py
class Solution:
    def minEnd(self, n: int, x: int) -> int:
        res = 0
        n -= 1

        x_bin = [0] * 64  # Binary representation of x
        n_bin = [0] * 64  # Binary representation of n-1

        for i in range(32):
            x_bin[i] = (x >> i) & 1
            n_bin[i] = (n >> i) & 1

        i_x = 0
        i_n = 0
        
        while i_x < 63:
            while i_x < 63 and x_bin[i_x] != 0:
                i_x += 1
            x_bin[i_x] = n_bin[i_n]
            i_x += 1
            i_n += 1

        for i in range(64):
            if x_bin[i] == 1:
                res += (1 << i)

        return res

File: 3133. Minimum Array End/solution_bitarray.cpp
#define ll long long

class Solution {
public:
    ll minEnd(int n, int x) {
        ll res = 0;
        n -= 1;

        vector<int> xBinary(64, 0);
        vector<int> nBinary(64, 0);

        ll longX = x;
        ll longN = n;

        for (int i = 0; i < 64; i++) {
            xBinary[i] = (longX >> i) & 1;
            nBinary[i] = (longN >> i) & 1;
        }

        int posX = 0, posN = 0;

        while (posX < 63) {
            while (xBinary[posX] != 0 && posX <63)
                posX++;
            
            xBinary[posX] = nBinary[posN];
            posX++;
            posN++;
        }

        for (int i = 0; i < 64; i++) {
            if (xBinary[i] == 1)
                res += pow(2, i);
        }

        return res;
    }
};
